# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/notification/NotificationParent.h
# Commit: b22268575af1
# Full Hash: b22268575af1492bb26867e00ba96bf98414212d
# Author: Kagami Sascha Rosylight <saschanaz@outlook.com>
# Date: 2025-04-01 09:40:15
# Description:
#   Bug 1956153 - Create actors directly on the main thread r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D243689
# ==============================================================================

diff -r 00dc9f4af854 -r b22268575af1 dom/notification/NotificationParent.h
--- a/dom/notification/NotificationParent.h	Mon Mar 31 19:49:03 2025 +0000
+++ b/dom/notification/NotificationParent.h	Mon Mar 31 19:53:58 2025 +0000
@@ -23,35 +23,34 @@
   Finished,
 };
 
+struct NotificationParentArgs {
+  NotNull<nsCOMPtr<nsIPrincipal>> mPrincipal;
+  NotNull<nsCOMPtr<nsIPrincipal>> mEffectiveStoragePrincipal;
+  bool mIsSecureContext;
+  nsString mScope;
+  IPCNotification mNotification;
+};
+
 class NotificationParent final : public PNotificationParent,
                                  public nsISupports,
                                  public SupportsWeakPtr {
   using IPCResult = mozilla::ipc::IPCResult;
 
  public:
-  // Threadsafe to pass the class from PBackground to main thread
-  NS_DECL_THREADSAFE_ISUPPORTS
-
-  NotificationParent(NotNull<nsIPrincipal*> aPrincipal,
-                     NotNull<nsIPrincipal*> aEffectiveStoragePrincipal,
-                     bool aIsSecureContext, const nsAString& aScope,
-                     const IPCNotification& aNotification)
-      : mPrincipal(aPrincipal),
-        mEffectiveStoragePrincipal(aEffectiveStoragePrincipal),
-        mIsSecureContext(aIsSecureContext),
-        mId(aNotification.id()),
-        mScope(aScope),
-        mOptions(aNotification.options()) {};
+  NS_DECL_ISUPPORTS
 
   nsresult HandleAlertTopic(AlertTopic aTopic);
   IPCResult RecvShow(ShowResolver&& aResolver);
   IPCResult RecvClose();
 
-  nsresult BindToMainThread(
+  static nsresult CreateOnMainThread(
+      NotificationParentArgs&& mArgs,
       Endpoint<PNotificationParent>&& aParentEndpoint,
       PBackgroundParent::CreateNotificationParentResolver&& aResolver);
 
  private:
+  explicit NotificationParent(NotificationParentArgs&& aArgs)
+      : mId(aArgs.mNotification.id()), mArgs(std::move(aArgs)) {};
   ~NotificationParent() = default;
 
   nsresult Show();
@@ -61,17 +60,10 @@
 
   Maybe<NotificationParent::ShowResolver> mResolver;
 
-  NotNull<nsCOMPtr<nsIPrincipal>> mPrincipal;
-  NotNull<nsCOMPtr<nsIPrincipal>> mEffectiveStoragePrincipal;
-  bool mIsSecureContext;
-
   // The ID generated by nsIAlertNotification during Show() function. It may
   // stay empty if the function fails.
   nsString mId;
-  nsString mScope;
-  const IPCNotificationOptions mOptions;
-
-  nsString mAlertName;
+  const NotificationParentArgs mArgs;
 
   // Whether it's now a dangling actor without corresponding OS notification,
   // either because it's closed or denied permission. We don't have to call